stumpy/stumpi.py

```diff
@@ -287,54 +314,64 @@ def _update(self, t):
 
         core.apply_exclusion_zone(D, D.shape[0] - 1, self._excl_zone, np.inf)
 
-        update_idx = np.argwhere(D[:l] < self._P[:l]).flatten()
-        self._I[update_idx] = l
-        self._P[update_idx] = D[update_idx]
-
-        I_last = np.argmin(D)
-        if np.isinf(D[I_last]):
-            I_new = np.append(self._I, -1)
-            P_new = np.append(self._P, np.inf)
-        else:
-            I_new = np.append(self._I, I_last)
-            P_new = np.append(self._P, D[I_last])
-        left_I_new = np.append(self._left_I, I_last)
-        left_P_new = np.append(self._left_P, D[I_last])
+        update_idx = np.argwhere(D[:l] < self._P[:l, -1]).flatten()
+        for i in update_idx:
+            idx = np.searchsorted(self._P[i], D[i], side="right")
+            core._shift_insert_at_index(self._P[i], idx, D[i])
+            core._shift_insert_at_index(self._I[i], idx, l)
+
+        # Calculating top-k matrix profile and (top-1) left matrix profile (and thier
+        # corresponding indices) for new subsequence whose distance profie is `D`
+        P_new = np.full(self._k, np.inf, dtype=np.float64)
+        I_new = np.full(self._k, -1, dtype=np.int64)
+        for i, d in enumerate(D):
+            if d < P_new[-1]:  # maximum value in sorted array P_new
+                idx = np.searchsorted(P_new, d, side="right")
+                core._shift_insert_at_index(P_new, idx, d)
+                core._shift_insert_at_index(I_new, idx, i)
+
+        left_I_new = I_new[0]
+        left_P_new = P_new[0]
 
         self._T = T_new
-        self._P = P_new
-        self._I = I_new
-        self._left_I = left_I_new
-        self._left_P = left_P_new
+        self._P = np.append(self._P, P_new.reshape(1, -1), axis=0)
+        self._I = np.append(self._I, I_new.reshape(1, -1), axis=0)
+        self._left_P = np.append(self._left_P, left_P_new)
+        self._left_I = np.append(self._left_I, left_I_new)
         self._QT = QT_new
         self._M_T = M_T_new
         self._Σ_T = Σ_T_new
 
     @property
     def P_(self):
         """
-        Get the matrix profile
+        Get the (top-k) matrix profile. When `k=1` (default), the first (and only)
+        column in this 2D array consists of the matrix profile. When `k > 1`, the
+        output has exactly `k` columns consisting of the top-k matrix profile.
         """
         return self._P.astype(np.float64)
```

Comment:
Sure. I will revise this docstring and others accordingly.

Reply:
@seanlaw 
So, I was wondering if we should change this to 1D? because, this is a private function and we can keep it 2D. 

If we want to keep it 1D, then should I still flatten `self.P_` (as you proposed in another comment) just to be on the safe side?

Reply:
@seanlaw 
Please ignore my previous comment. I got confused and made a mistake. Sorry.